client_src/js/printerManager/printer-constructor.js

```diff
@@ -0,0 +1,323 @@
+    }
+    let currentIndex = JSON.parse(JSON.stringify(newPrintersIndex));
+    document
+      .getElementById(`delButton-${newPrintersIndex}`)
+      .addEventListener("click", (event) => {
+        removeLine(document.getElementById(`newPrinterCard-${currentIndex}`));
+        const table = document.getElementById("printerNewTable");
+        if (table.rows.length === 1) {
+          if (!table.classList.contains("d-none")) {
+            table.classList.add("d-none");
+          }
+        }
+      });
+    document
+      .getElementById(`saveButton-${newPrintersIndex}`)
+      .addEventListener("click", (event) => {
+        PrintersManagement.savePrinter(event.target);
+      });
+    newPrintersIndex++;
+  }
+
+  static async importPrinters() {
+    return function (e) {
+      const theBytes = e.target.result; // .split('base64,')[1];
+      // Initial JSON validation
+      if (Validate.JSON(theBytes)) {
+        // If we can parse the file.
+
+        // Grab uploaded file contents into an object
+        const importPrinters = JSON.parse(theBytes);
+        // Loop over import only importing printers with correct fields.
+        for (let index = 0; index < importPrinters.length; index++) {
+          const printer = {
+            printerURL: "Key not found",
+            cameraURL: "Key not found",
+            name: "Key not found",
+            group: "Key not found",
+            apikey: "Key not found"
+          };
+          if (typeof importPrinters[index].name !== "undefined") {
+            printer.name = importPrinters[index].name;
+          }
+          if (typeof importPrinters[index].printerURL !== "undefined") {
+            printer.printerURL = importPrinters[index].printerURL;
+          }
+          if (typeof importPrinters[index].cameraURL !== "undefined") {
+            printer.cameraURL = importPrinters[index].cameraURL;
+          }
+          if (typeof importPrinters[index].group !== "undefined") {
+            printer.group = importPrinters[index].group;
+          }
+          if (typeof importPrinters[index].apikey !== "undefined") {
+            printer.apikey = importPrinters[index].apikey;
+          }
+          PrintersManagement.addPrinter(printer);
+        }
+        UI.createAlert(
+          "success",
+          "Successfully imported your printer list, Please check it over and save when ready.",
+          3000
+        );
+      } else {
+        UI.createAlert(
+          "error",
+          "The file you have tried to upload contains json syntax errors.",
+          3000
+        );
+      }
+    };
+  }
+
+  static async deletePrinter(deletedPrinters) {
+    if (deletedPrinters.length > 0) {
+      try {
+        const printersToRemove = await OctoFarmClient.post(
+          "printers/remove",
+          deletedPrinters
+        );
+        // Should help with SSE event re-building the deleted printer causing a refresh to be needed to actually clear from UI
+        await UI.delay(5000);
```

Comment:
We should avoid this at **all cost**. If we have problems with delayed data then fix the data stream, not the original point where it's requested

This is a huge red flag especially 3000 sec

Reply:
5 seconds not 3000

also outside of the scope of this PR

Reply:
I meant to write 3000ms. Thats still wrong on my side! 😄 

5 seconds is a long time and there's no cancellation. All I can see it's introduced in this PR, so I feel the obligation to review it here as well.

Reply:
It's a patch work to the current method of SSE events. When the delete function is fired, this function get's the acceptance response from the server side code, before the printerClean function has actually cleaned out the removed printer(s). Fixing the data stream requires a lot more work than is contained within this PR, because there's nothing within the backend to know about the printerClean function successfully running from the API endpoint. The delay doesn't lock up the client, and it also does nothing to the delete function been fired, it just delays the actual removal of the element and the UI notification trigger for the user.